man/man3/menuhit.3
.\" 
.\"									
.\"	Copyright (c) 1987,1988,1989,1990,1991,1992   AT&T		
.\"			All Rights Reserved				
.\"									
.\"	  THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF AT&T.		
.\"	    The copyright notice above does not evidence any		
.\"	   actual or intended publication of such source code.		
.\"									
.\" 
.ds ZZ APPLICATION DEVELOPMENT PACKAGE
.TH MENUHIT 3R
.XE "menuhit()"
.SH NAME
menuhit \- present user with menu and get selection
.SH SYNOPSIS
.ft B
int menuhit (m, n) 
.sp
Menu *m;  
.br
int n;
.PP
.nf
\s-1
.ft CM
typedef struct Menu {
  char  **item;       /* string array, ending with 0 */
  short prevhit;       /* retained from previous call */
  short prevtop;       /* retained from previous call */
  char  *(*generator)(); /* used if item == 0 */
} Menu;
.ft R
.fi
\s+1
.SH DESCRIPTION
The
.I menuhit
function
presents the user with a menu specified by the Menu pointer
.I m
and returns an integer indicating the selection made,
or -1 for no selection.
The
.I n
argument
specifies which button to use for the interaction: 1, 2 or 3.
The
.I menuhit
function
assumes that the button is already depressed when it is called.
The user makes a selection by lifting the button when the cursor
points to the desired selection.
Lifting the button outside the menu indicates no selection.
.PP
The maximum number of menu items displayed at any one time is sixteen items.
When the number of items is 16 or less,
all of the items are displayed and are centered one entry per line in the menu.
This is the normal menu mode.
When there are more than 16 menu items to be displayed, the menu becomes a 
.B scrolling
menu.
The left portion of the menu contains a scroll bar which is 
used for scrolling quickly through the menu selections.
The vertical size of the scroll bar is an indication of the size
of the user's view of the menu (16 items) relative to the number
of selections in the entire menu.  
.PP
There are two ways to scroll through the menu items.
The first is to move the mouse cursor to the left side of the menu
into the scroll bar area.
By moving the mouse cursor up or down within the scroll bar area the menu items
will scroll accordingly.
The second method used to scroll through the menu 
items is to place the mouse cursor on the top or bottom of the menu
list.  The menu will scroll up or down by one item at a time if 
there are additional items to be displayed in that direction.
.PP
Menus may be generated dynamically from a program by initializing the generator 
function specified in the Menu structure.
If
.I item
is set to 0 when \f2menuhit\f1 is called,
then the routine specified by generator
is called with one parameter which is an integer index beginning at 0.
The generator must return a pointer to a character string containing the text
for the menu selection.
This generator function is called repeatedly 
with the index increasing by 1 until the generator returns a \s-1NULL\s+1,
indicating the end of the menu selections.
The generator function is called
each time that the \f2menuhit\f1 routine is called with item set to 0 or \s-1NULL\s+1.
.PP
Another facility provided by
.I menuhit
is the notion of a spread character.
A spread character is any ascii character with the high-order bit set.
The spread character acts somewhat like a spring pushing against the text
and borders within a menu entry.
The spread character can be placed at the beginning, middle, or end
of the string defining the menu entry.
If placed at the beginning of the string,
the text in the menu item will be right-justified.
If placed at the end of the string,
the text will be left justified.
If placed in the middle of the string, the text on each
side of the spread character will be pushed against the menu border.
In each of the cases,
the space created by the spread character will be filled in with
the ascii character contained in the spread character.
For entries without a spread character,
the default is to have the text centered.
.PP
Whenever a menu is displayed, the original screen image obscured by the menu is
saved in the terminal and then later restored when the menu disappears.
If the terminal is
out of memory and therefore cannot save the screen image,
then the menu will be displayed in \s-1XOR\s+1 (exclusive or) mode on top of the existing screen image.
Menu items may still be selected in this mode but the items might be hard to read.
To remedy this problem,
memory must be freed up by either deleting or
reshaping \f2layers\f1 before the menu is displayed.
.SH EXAMPLE
The following example includes both a menu with the spread character
and a menu that is dynamically generated.
.PP
.RS 0
\s-1
.nf
.ft CM
#include <dmd.h>

char *menutext[] = {
	"left\e240",           /* space char with
	                   high bit set */
	"\e256right",          /* . char with high
	                          bit set */
	"middle",
	"left\e337right",      /* _ char with high
	                          bit set */
	"a very long string",
	NULL };
Menu menu = { menutext };      /* static menu */

/*	Note the above menu will appear as:

       --------------------
       |left              |
       |.............right|
       |      middle      |
       |left_________right|
       |a very long string|
       --------------------
*/

static char digits[50];
static int index;
#define MAXDIGIT 25

pic9( n )
long n;
{
	if(n/10) pic9 (n/10);
	digits[index++] = "0123456789"[n%10];
}
.bp
char *
decimal(i)
long i;
{
	index = 0;
	pic9 ((long)i);
	digits[index] = '\e0';
	return digits;
}
char *
generate(i)
int i;
{
	if( i > MAXDIGIT ) return NULL;
	return decimal (i);
}

Menu menugen = {0, 0, 0, generate};
            /* dynamically generated menu */


main()
{
	int m;

	request (MOUSE);
	while( !bttn3() )
		wait (MOUSE);
	m = menuhit (&menu, 3);
	if ( m == 0 )
		exit();
	while( !bttn2() )
		wait (MOUSE);
	m = menuhit (&menugen, 2);
}
\fR
.fi
.RE
\s+1
.SH SEE ALSO
layers(1).


dmdebug/term/menuhit.c
/*	Copyright (c) 1984 AT&T	*/
/*	  All Rights Reserved  	*/

/*	THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF AT&T	*/
/*	The copyright notice above does not evidence any   	*/
/*	actual or intended publication of such source code.	*/

/*
#include <sccsid.h>
VERSION(@(#)menuhit.c	2.4);
*/

#include "dmd.h"
#include "font.h"

typedef struct NewMenu{
	char	**item;			/* string array, ending with 0 */
	short	prevhit;		/* retained from previous call */
	short	prevtop;		/* ditto */
	char	*(*generator)();	/* used if item == 0 */
} NewMenu;

#define scale(x, inmin, inmax, outmin, outmax)\
	(outmin + muldiv(x-inmin,outmax-outmin,inmax-inmin))

#define bound(x, low, high) min(high, max( low, x ))

#define SPACING		14
#define DISPLAY		16
#define CHARWIDTH	9
#define DELTA		6
#define BARWIDTH	18

static Bitmap dphysical = { (Word *) 0x700000, 25, 0, 0, XMAX, YMAX };

static char **table;
static char *
tablegen(i)
{
	return table[i];
}

smenuhit(m, but)
register NewMenu *m;
{
	register int width, i, j, top, newtop, hit, newhit, items, lines, length;
	Point p, q, savep, baro, barc;
	Rectangle sr, tr, mr;	/* scroll, text, menu */
	register Bitmap *b;
	register char *s, *(*generator)(), *from, *to, fill[64];

#define sro sr.origin
#define src sr.corner
#define tro tr.origin
#define trc tr.corner
#define mro mr.origin
#define mrc mr.corner

	generator = (table=m->item) ? tablegen : m->generator;
	p = mouse.xy;
	for(length = items = 0; s=(*generator)(items); ++items)
		length = max(length, strlen(s));
	if(items == 0)
		return -1;
	width = length*CHARWIDTH+10;
	sro.x = sro.y = src.x = tro.x = mro.x = mro.y = 0;
	if(items <= DISPLAY)
		lines = items;
	else{
		lines = DISPLAY;
		tro.x = src.x = BARWIDTH;
		sro.x = sro.y = 1;
	}
#define ASCEND 2
	tro.y = ASCEND;
	mrc = trc = add(tro, Pt(width, min(items, lines)*SPACING));
	src.y = mrc.y-1;
	newtop = bound(m->prevtop, 0, items-lines);
	p.y -= bound(m->prevhit, 0, lines-1)*SPACING+SPACING/2;
	p.x = bound(p.x-(src.x+width/2), 0, XMAX-mrc.x);
	p.y = bound(p.y, 0, YMAX-mrc.y);
	sr = raddp(sr, p);
	tr = raddp(tr, p);
	mr = raddp(mr, p);
	b = balloc(mr);
	cursinhibit();
	if(b)
		bitblt(&dphysical, mr, b, mro, F_STORE);
	rectf(&dphysical, mr, F_OR);
	cursallow();
PaintMenu:
	cursinhibit();
	rectf(&dphysical, inset(mr, 1), F_CLR);
	cursallow();
	top = newtop;
	if(items > DISPLAY){
		baro.y = scale(top, 0, items, sro.y, src.y);
		baro.x = sr.origin.x;
		barc.y = scale(top+DISPLAY, 0, items, sro.y, src.y);
		barc.x = sr.corner.x;
		rectf(&dphysical, Rpt(baro,barc), F_XOR);
	}
	for(p=tro, i=top; i < min(top+lines, items); ++i){
		q = p;
		from = generator(i);
		for(to = &fill[0]; *from; ++from)
			if(*from & 0x80)
				for(j=length-(strlen(from+1)+(to-&fill[0])); j-->0;)
					*to++ = *from & 0x7F;
			else
				*to++ = *from;
		*to = '\0';
		q.x += (width-jstrwidth(fill))/2;
		string(&defont, fill, &dphysical, q, F_XOR);
		p.y += SPACING;
	}
	savep = mouse.xy;
	for(newhit = hit = -1; bttn(but); nap(2)){
		if(ptinrect(p = mouse.xy, sr)){
			if(ptinrect(savep,tr)){
				p.y = (baro.y+barc.y)/2;
				cursset(p);
			}
			newtop = scale(p.y, sro.y, src.y, 0, items);
			newtop = bound(newtop-DISPLAY/2, 0, items-DISPLAY);
			if(newtop != top)
/* ->->-> */			goto PaintMenu;
		}else if(ptinrect(savep,sr)){
			register dx, dy;
			if(abs(dx = p.x-savep.x) < DELTA)
				dx = 0;
			if(abs(dy = p.y-savep.y) < DELTA)
				dy = 0;
			if(abs(dy) >= abs(dx))
				dx = 0;
			else
				dy = 0;
			cursset(p = add(savep, Pt(dx,dy)));
		}
		savep = p;
		newhit = -1;
		if(ptinrect(p, tr)){
			newhit = bound((p.y-tro.y)/SPACING, 0, lines-1);
			if(newhit!=hit && hit>=0
			 && abs(tro.y+SPACING*newhit+SPACING/2-p.y) > SPACING/3)
				newhit = hit;
		}
		if(newhit != hit){
			flip(tr, hit);
			flip(tr, hit = newhit);
		}
		if(newhit==0 && top>0){
			newtop = top-1;
			p.y += SPACING;
			cursset(p);
/* ->->-> */		goto PaintMenu;
		}
		if(newhit==DISPLAY-1 && top<items-lines){
			newtop = top+1;
			p.y -= SPACING;
			cursset(p);
/* ->->-> */		goto PaintMenu;
		}
	}
	if(b){
		cursinhibit();
		screenswap(b, b->rect, b->rect);
		cursallow();
		bfree(b);
	}
	if(hit>=0){
		m->prevhit = hit;
		m->prevtop = top;
		return hit+top;
	}else
		return -1;
}

static
flip(r,n)
	Rectangle r;
{
	if(n<0)
		return;
	++r.origin.x;
	r.corner.y = (r.origin.y += SPACING*n-1) + SPACING;
	--r.corner.x;
	rectf(&dphysical, r, F_XOR);
}


